docs/v1/P4Runtime-Spec.mdk

```diff
@@ -706,9 +712,15 @@ read state from the data plane, among others described later. P4Runtime relies
 on clients identifying themselves in every write request, by including the
 values `device_id`, `role_id`, and `election_id` in all write requests. The
 server trusts clients not to use a triple of values other than their own in
-their write requests. gRPC provides authentication methods [@gRPCAuth] that can
-be deployed to prevent untrusted clients from creating channels, and thus from
-making changes or even reading the state of the server.
+their write requests. gRPC provides authentication methods [@gRPCAuth] that
+should be deployed to prevent untrusted clients from creating channels, and thus
+from making changes or even reading the state of the server.
+
+To avoid misconfiguration and abusing identifiers even after authenticating the
+gRPC channel, the controller should authenticate the P4Runtime server's
+`device_id`. Furthermore, the P4Runtime server should be protected from
```

Comment:
could you clarify what you mean by "authenticate the P4Runtime server's `device_id`" and why it is important?

Reply:
If the controller issues both the `StreamMessageRequest` and the `SetForwardingPipelineConfigRequest`, how can the device ids be different?
The controller is in charge of building the `SetForwardingPipelineConfigRequest` message from the `ForwardingPipelineConfig` message sent by the "P4 build machine".

Reply:
I agree with you that the`StreamMessageRequest` and the `SetForwardingPipelineConfigRequest` will have the same device ids as they are from the controller.
However, my suggestion is to have the controller ask the switch for it's `device_id`, and then check if `device_id` the switch sends is the one the controller expects. Does that make sense?

Reply:
Not sure I follow. The switch doesn't just choose its device id, it is assigned by the controller or a separate entity.
 Basically I would imagine the controller always has access to this information for every device from the get go:
(device_id, switch IP , switch port, pub key)
One the secure gRPC channel is created, summing that somehow there is a device_id mismatch, the controller would get an error status: INVALID_ARGUMENT, "Unknown device_id"
You cannot ask the switch for the device id BTW, because several devices may be on the switch, with a single P4Runtime service.
Maybe the issue is that I'm not sure about what point you are trying to make here, besides "the Forwarding Pipeline Config should be delivered to the switch in a secure fashion". Since the release is supposed to happen very soon (it was supposed to happen today), maybe we should have a quick call to sync-up, or only keep the section for which we are on the same page, for this release of P4Runtime.